### PR TITLE
feat: Implement DAP protocol in Nargo

### PR DESCRIPTION
**Pull request by ggiraldez**
# Description

Implements the DAP protocol to allow a tool such as VS.Code to drive the debugger in a Noir binary project. A separate PR for the [VS.Code Noir extension](https://github.com/noir-lang/vscode-noir) will be submitted later.

## Problem

Part of #3015 

The beginning of this implementation was heavily inspired by @dmvict [implementation](https://github.com/noir-lang/noir/issues/3094) of an alternative debugger for Noir programs.

## Summary

This PR implements a new `nargo` subcommand `dap`. This starts a DAP server in single session mode through stdin/stdout and waits for a launch request. Using the arguments in the launch request (`projectFolder`, and optionally `package` and `proverName`) it compiles the Noir binary package and starts it in debug mode. Through DAP requests, a tool can then step through the program, set breakpoints and list the generated opcodes using a disassemble request.

## Documentation

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [X] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by TomAFrench — COMMENTED** [review]
Testing functionality atm, raising a few surface things from an initial look at the code.

File: tooling/debugger/src/dap.rs [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+            let Ok(line_index) = &simple_files[&file_id].line_index((), span.start() as usize) else {
+                return;
+            };
+            let line_number = line_index + 1;
+
+            if let Entry::Vacant(e) = result.entry(file_id) {
+                e.insert(vec![(line_number, *opcode_location)]);
+            } else {
+                result.get_mut(&file_id).unwrap().push((line_number, *opcode_location));
+            }
```
[the remark this answers is not in the document]
Comment (ggiraldez) [reply tooling/debugger/src/dap.rs:109]: Nice! I spent quite a bit of time looking into how to do this idiomatically in Rust, but couldn't find a good answer. Thanks!

File: tooling/debugger/src/dap.rs [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+                    match granularity {
+                        SteppingGranularity::Instruction => self.handle_step(req)?,
+                        _ => self.handle_next(req)?,
+                    }
+                }
+                Command::Continue(_) => {
+                    self.handle_continue(req)?;
+                }
+                Command::Scopes(_) => {
+                    // FIXME
```
[the remark this answers is not in the document]
Comment (ggiraldez) [reply tooling/debugger/src/dap.rs:201]: Yeah, that needs an actual implementation. We have something working already at https://github.com/manastech/noir/blob/dap-with-vars/tooling/debugger/src/dap.rs#L550 This was meant as a placeholder. I'll clarify with a proper comment.

**Review by TomAFrench — APPROVED** [review]
I've tested this and got it working in VSCode now. There's a couple of things we could probably handle better, e.g. 

- stepping into a function results in the debugger sending me to the relevant line in `main` rather than the deepest function which is what I would expect.
- better automatic finding of the project root

These can get handled later however. Thanks for this and apologies for the delay!

**Comment by TomAFrench** [comment]
Argh, sorry looks like there are some merge conflicts that have arisen.

**Comment by ggiraldez** [comment]
> Argh, sorry looks like there are some merge conflicts that have arisen.

No worries, I'll update to latest master.

**Pull request merged**